Tools to work with PULL_REFS

csrwng opened the pull request:
Add a release-ci command that provides tooling to work with PULL_REFS (for now)

csrwng commented:
@stevekuznetsov ptal ... not sure you'll like that by adding cobra I'm vendoring the internet, but I'd rather do that now if in the not so distant future we'll be adding more commands.

stevekuznetsov commented on tools/release-ci/cmd/cloneref.go:
```diff
@@ -0,0 +1,89 @@
+
+// clonerefCmd clones a PULL_REFS specification
+var clonerefCmd = &cobra.Command{
+	Use:   "cloneref REPOSITORY_URL [PULL_REFS]",
```
We just expect to call this for every repo?

csrwng replied:
yes for every PR we want to clone

stevekuznetsov commented on tools/release-ci/cmd/cloneref.go:
```diff
@@ -0,0 +1,89 @@
+	
+PULL_REFS is expected to be in the form of:
+
+base_branch:commit_sha_of_base_branch,pull_request_no:commit_sha_of_pull_request_no,...
```
Upstream allows branch name not to have SHA -- we need to determine if we are going to interpret "build this with latest Origin RPMs" by build-time resolving the current stable SHA for Origin artifacts to get that info into this

csrwng replied:
This code handles not having a sha for the branch

stevekuznetsov replied:
Hmm, sorry -- what i meant to say is we still probably want to know what SHA we are working with when we say "latest" -- otherwise it's going to be super hard to determine what version of X some PR was built with. So before this script is called we'd want to resolve that?

csrwng replied:
got it

stevekuznetsov commented on tools/release-ci/cmd/cloneref.go:
```diff
@@ -0,0 +1,89 @@
+And for multiple pull requests that have been batched:
+master:97d901d,4:bcb00a1,6:ddk2tka
+
+If PULL_REFS is not specified as an argument, the command will try to find it as
```
@kargakis how is multi-pull-refs passed? one envar?

csrwng replied:
it is still one param (from what I see in jenkins jobs)

stevekuznetsov replied:
I don't think any of ours right now are doing multi-repo

stevekuznetsov commented on tools/release-ci/cmd/cloneref.go:
```diff
@@ -0,0 +1,89 @@
+	}
+
+	if len(args) == 0 {
+		return fmt.Errorf("A repository URL must be specified as an argument")
```
I'm pretty sure this is just host/org/name  upstream in `--repo=repo-name=branch-name:sha,pull-id:sha,..." so `github.com/openshift/origin` -- unclear if you expect there to be more to this URL or not

csrwng replied:
I just need a git url whatever shape that takes

stevekuznetsov commented on tools/release-ci/cmd/root.go:
```diff
@@ -0,0 +1,59 @@
+var RootCmd = &cobra.Command{
+	Use:   "release-ci",
+	Short: "tools for OpenShift release pipelines",
+	Long: `This rool is used by OpenShift release pipelines to clone repository
```
tool

stevekuznetsov commented on tools/release-ci/glide.lock:
```diff
@@ -0,0 +1,52 @@
+hash: 4936647ed332b0acaca71ad115ebb9b3ac2c8b791ccc3d2531c8ede1e19d7eb4
```
Why `glide` over `dep`? is `dep` usable?

csrwng commented:
slimmed down vendor dir, removed binary from commit

stevekuznetsov commented on tools/release-ci/pkg/pullrefs/cloneref.go:
```diff
@@ -0,0 +1,92 @@
+	}
+
+	if !existing {
+		runner.Add("mkdir", "-p", dir)
```
Why not do this in Go and have runner be 

```go
type Step func() error

type Runner interface {
    Add(step Step)
    Run() error
}
``` 
?

csrwng replied:
sure... can make it a func ... all I wanted was something to let me make the code clearer so I don't have a lot of error checking cluttering things.

stevekuznetsov replied:
Yeah, makes sense. Just wanted to keep the shelling out to a minimum

csrwng commented:
thanks @stevekuznetsov addressed comments so far. For multirepo jobs I think we can still invoke the tool as many times as we need for each one. Also, what do you think about the location for this? should we have a single place where we put all go code?

stevekuznetsov commented:
Yeah, makes sense to have e.g. `pkg/logging/` for my stuff and `pkg/checkout/` or whatever for this, we can have top-level `cmd/` entries for compiled entrypoints, and share one `vendor/` folder? @smarterclayton does this still seem like stuff we want in this repo, or should we have our own?

smarterclayton commented:
I think it's ok for common tools here for the release process, or things reused by other projects.  However, keep in mind that anything that requires running go requires compiling go, so we should be judicious and at a certain point an external project may make sense.

We should move the release images into this project.

stevekuznetsov commented:
ack
